File: tests/snapshot-recovery.test.ts
import { describe, expect, it } from "vitest";
import type { SyncCodec } from "../src/core/types.js";
import {
  createV1EnvelopeCrypto,
  createWebCryptoBackend,
  defineV1CompatibilityProfile,
  deriveContentKey,
  generateRecoveryCode,
  type SyncEnvelopeV1,
  type V1KeyMetadata,
} from "../src/crypto/index.js";
import { createSnapshotSync } from "../src/snapshot/index.js";

type Value = { items: string[]; updatedAt: string };

const backend = createWebCryptoBackend();
const codec: SyncCodec<Value> = {
  serialize: (value) => value,
  parse: (value) => value as Value,
  merge: (local, remote) => ({
    items: [...new Set([...remote.items, ...local.items])].sort(),
    updatedAt: local.updatedAt > remote.updatedAt ? local.updatedAt : remote.updatedAt,
  }),
  fingerprint: (value) => JSON.stringify([...value.items].sort()),
  updatedAt: (value) => value.updatedAt,
};

function profile(versions: { readVersions?: readonly (1 | 2)[]; writeVersion?: 1 | 2 } = {}) {
  return defineV1CompatibilityProfile({
    appId: "recovery-app",
    filename: "recovery-app.json",
    aad: "recovery-app-v1",
    hkdfInfo: "recovery-app content key",
    compression: "gzip-if-smaller",
    passkey: {
      rpName: "Recovery App",
      userName: "user",
      userDisplayName: "User",
      algorithm: -7,
      residentKey: "required",
      userVerification: "required",
      timeoutMs: 60_000,
    },
    ...versions,
  });
}

const V2 = { readVersions: [1, 2] as const };

/**
 * A stand-in for a passkey provider: each credential's PRF secret lives in
 * `authenticator`, and keys derive from it exactly as the real providers do.
 * Deleting a credential simulates losing the passkey.
 */
function passkeys(appProfile: ReturnType<typeof profile>) {
  const authenticator = new Map<string, Uint8Array>();
  let created = 0;
  return {
    authenticator,
    provider: {
      async create() {
        const credentialId = `credential-${++created}`;
        const secret = backend.randomBytes(32);
        authenticator.set(credentialId, secret);
        const metadata: V1KeyMetadata = {
          credentialId,
          rpId: "recovery.example",
          prfInput: backend.randomBytes(32),
          kdfSalt: backend.randomBytes(32),
        };
        return { metadata, key: await deriveContentKey(appProfile, secret, metadata.kdfSalt, backend) };
      },
      async unlock(envelope: SyncEnvelopeV1) {
        const secret = authenticator.get(envelope.credentialId);
        if (!secret) throw new Error("This passkey is not available on this device.");
        return deriveContentKey(
          appProfile,
          secret,
          Uint8Array.from(Buffer.from(envelope.kdfSalt, "base64url")),
          backend,
        );
      },
      clear: () => undefined,
    },
  };
}

function device(
  appProfile: ReturnType<typeof profile>,
  cloud: { envelope: SyncEnvelopeV1 | null },
  keys: ReturnType<typeof passkeys>,
  local: Value,
) {
  const state = { local };
  const envelopeCrypto = createV1EnvelopeCrypto(appProfile, codec, backend);
  const controller = createSnapshotSync({
    appId: appProfile.appId,
    codec,
    envelopeCrypto,
    keyProvider: keys.provider,
    authorizationProvider: { authorize: async () => ({ accessToken: "token" }), clear: () => undefined },
    cloudStore: {
      find: async () => (cloud.envelope ? { fileId: "file", envelope: cloud.envelope } : null),
      write: async (_appId, envelope) => {
        cloud.envelope = envelope;
        return "file";
      },
    },
    readLocal: () => state.local,
    applyMerged: (value) => {
      state.local = value;
    },
    envelopeUpdatedAt: (envelope) => envelope.updatedAt,
  });
  return { controller, envelopeCrypto, state };
}

/** The snapshot in the cloud, failing loudly if there is none. */
function stored(cloud: { envelope: SyncEnvelopeV1 | null }): SyncEnvelopeV1 {
  if (!cloud.envelope) throw new Error("No snapshot has been written.");
  return cloud.envelope;
}

const at = (minute: number) => `2026-09-23T00:${String(minute).padStart(2, "0")}:00.000Z`;

describe("snapshot v2 and recovery codes", () => {
  it("leaves v1 exactly as it was unless a profile opts in", async () => {
    const appProfile = profile();
    const cloud = { envelope: null as SyncEnvelopeV1 | null };
    const phone = device(appProfile, cloud, passkeys(appProfile), { items: ["a"], updatedAt: at(1) });
    await phone.controller.setup();
    expect(cloud.envelope).toMatchObject({ schemaVersion: 1 });
    expect(Object.keys(stored(cloud)).sort()).toEqual(
      ["algorithm", "ciphertext", "credentialId", "kdfSalt", "nonce", "prfInput", "rpId", "schemaVersion", "updatedAt"]
        .concat(stored(cloud).compression ? ["compression"] : [])
        .sort(),
    );
    await expect(phone.controller.setRecoveryCode(await generateRecoveryCode(backend))).rejects.toMatchObject({
      code: "configuration",
    });
  });

  it("refuses a v2 snapshot on a device whose profile does not read v2", async () => {
    const v2Profile = profile({ ...V2, writeVersion: 2 });
    const cloud = { envelope: null as SyncEnvelopeV1 | null };
    const keys = passkeys(v2Profile);
    await device(v2Profile, cloud, keys, { items: ["a"], updatedAt: at(1) }).controller.setup();
    expect(cloud.envelope?.schemaVersion).toBe(2);
    const old = device(profile(), cloud, keys, { items: [], updatedAt: at(0) });
    await expect(old.controller.sync("foreground")).rejects.toThrow(/readVersions/);
  });

  it("recovers a lost passkey on a new device with the code alone", async () => {
    const appProfile = profile(V2);
    const cloud = { envelope: null as SyncEnvelopeV1 | null };
    const keys = passkeys(appProfile);
    const phone = device(appProfile, cloud, keys, { items: ["a"], updatedAt: at(1) });
    await phone.controller.setup();
    const code = await generateRecoveryCode(backend);
    await phone.controller.setRecoveryCode(code);
    expect(cloud.envelope).toMatchObject({ schemaVersion: 2, appId: "recovery-app" });
    expect(cloud.envelope?.recoveryKey).toBeDefined();

    // Ordinary sync keeps the version and the recovery lock untouched.
    phone.state.local = { items: ["a", "b"], updatedAt: at(2) };
    const lockBefore = cloud.envelope?.recoveryKey;
    await phone.controller.sync("change");
    expect(cloud.envelope?.schemaVersion).toBe(2);
    expect(cloud.envelope?.recoveryKey).toEqual(lockBefore);

    // The passkey is gone. A new device has only the code.
    keys.authenticator.clear();
    const newDevice = device(appProfile, cloud, keys, { items: ["c"], updatedAt: at(3) });
    await expect(newDevice.controller.sync("foreground")).rejects.toThrow(/not available/);
    await expect(newDevice.controller.recover(code)).resolves.toMatchObject({
      operation: "recover",
      outcome: "recovered",
      value: { items: ["a", "b", "c"] },
    });
    // Locked under the new passkey from here on, and the code still works.
    await expect(newDevice.controller.sync("foreground")).resolves.toMatchObject({ outcome: "unchanged" });
    await expect(newDevice.envelopeCrypto.decryptWithRecoveryCode(stored(cloud), code)).resolves.toMatchObject({
      items: ["a", "b", "c"],
    });
  });

  it("distinguishes a typo from a wrong code, and removes the code on request", async () => {
    const appProfile = profile(V2);
    const cloud = { envelope: null as SyncEnvelopeV1 | null };
    const phone = device(appProfile, cloud, passkeys(appProfile), { items: ["a"], updatedAt: at(1) });
    await phone.controller.setup();
    const code = await generateRecoveryCode(backend);
    await phone.controller.setRecoveryCode(code);
    const typo = code.slice(0, 5) + (code[5] === "A" ? "B" : "A") + code.slice(6);
    await expect(phone.envelopeCrypto.decryptWithRecoveryCode(stored(cloud), typo)).rejects.toThrow(/typo/);
    await expect(
      phone.envelopeCrypto.decryptWithRecoveryCode(stored(cloud), await generateRecoveryCode(backend)),
    ).rejects.toThrow(/does not unlock/);
    await phone.controller.setRecoveryCode(null);
    expect(cloud.envelope?.recoveryKey).toBeUndefined();
    await expect(phone.envelopeCrypto.decryptWithRecoveryCode(stored(cloud), code)).rejects.toThrow(
      /no recovery code/,
    );
  });

  it("migrates explicitly and reversibly, and moving to v1 drops the code", async () => {
    const appProfile = profile(V2);
    const cloud = { envelope: null as SyncEnvelopeV1 | null };
    const keys = passkeys(appProfile);
    const phone = device(appProfile, cloud, keys, { items: ["a"], updatedAt: at(1) });
    await phone.controller.setup();
    await phone.controller.setRecoveryCode(await generateRecoveryCode(backend));
    await phone.controller.migrateVersion(1);
    expect(cloud.envelope).toMatchObject({ schemaVersion: 1 });
    expect(cloud.envelope?.recoveryKey).toBeUndefined();
    // A device that only reads v1 opens it again.
    const old = device(profile(), cloud, keys, { items: [], updatedAt: at(0) });
    await expect(old.controller.sync("foreground")).resolves.toMatchObject({ value: { items: ["a"] } });
    await phone.controller.migrateVersion(2);
    expect(cloud.envelope?.schemaVersion).toBe(2);
  });

  it("authenticates the whole v2 header and its appId", async () => {
    const appProfile = profile({ ...V2, writeVersion: 2 });
    const cloud = { envelope: null as SyncEnvelopeV1 | null };
    const keys = passkeys(appProfile);
    const phone = device(appProfile, cloud, keys, { items: ["a"], updatedAt: at(1) });
    await phone.controller.setup();
    const envelope = stored(cloud);
    const key = await keys.provider.unlock(envelope);
    await expect(phone.envelopeCrypto.decrypt({ ...envelope, updatedAt: at(9) }, key)).rejects.toMatchObject({
      code: "crypto",
    });
    const other = createV1EnvelopeCrypto({ ...appProfile, appId: "another-app" }, codec, backend);
    await expect(other.decrypt(envelope, key)).rejects.toThrow(/belongs to recovery-app/);
  });

  it("validates versions in the profile", () => {
    expect(() => profile({ readVersions: [2] })).toThrow(/must include 1/);
    expect(() => profile({ writeVersion: 2 })).toThrow(/one of readVersions/);
  });
});

File: android/synckit/src/test/java/com/keyneom/synckit/snapshot/SnapshotRecoveryTest.kt
package com.keyneom.synckit.snapshot

import android.app.Activity
import com.keyneom.synckit.core.Authorization
import com.keyneom.synckit.core.AuthorizationProvider
import com.keyneom.synckit.core.CloudStore
import com.keyneom.synckit.core.CreatedKey
import com.keyneom.synckit.core.KeyProvider
import com.keyneom.synckit.core.SnapshotOperation
import com.keyneom.synckit.core.StoredEnvelope
import com.keyneom.synckit.core.SyncCodec
import com.keyneom.synckit.core.SyncKitError
import com.keyneom.synckit.core.SyncKitErrorCode
import com.keyneom.synckit.core.SyncOutcome
import com.keyneom.synckit.core.SyncReason
import com.keyneom.synckit.crypto.Base64Url
import com.keyneom.synckit.crypto.PasskeyProfile
import com.keyneom.synckit.crypto.RecoveryCodes
import com.keyneom.synckit.crypto.SyncEnvelopeV1
import com.keyneom.synckit.crypto.SyncKitJson
import com.keyneom.synckit.crypto.V1CompatibilityProfile
import com.keyneom.synckit.crypto.V1Compression
import com.keyneom.synckit.crypto.V1EnvelopeCrypto
import com.keyneom.synckit.crypto.V1KeyMetadata
import java.io.File
import java.security.SecureRandom
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class SnapshotRecoveryTest {
    private val random = SecureRandom()
    private fun bytes(length: Int) = ByteArray(length).also(random::nextBytes)

    private fun profile(
        appId: String = "recovery-app",
        readVersions: List<Int> = listOf(1),
        writeVersion: Int = 1,
        aad: String = "recovery-app-v1",
        hkdfInfo: String = "recovery-app content key",
    ) = V1CompatibilityProfile(
        appId = appId,
        filename = "$appId.json",
        aad = aad,
        hkdfInfo = hkdfInfo,
        compression = V1Compression.GZIP_IF_SMALLER,
        passkey = PasskeyProfile("Recovery App", "user", "User"),
        readVersions = readVersions,
        writeVersion = writeVersion,
    )

    /** JSON `{items, updatedAt}`, merged as a sorted union — the same shape the web tests use. */
    private val codec = object : SyncCodec<JsonObject> {
        override fun serialize(value: JsonObject): ByteArray = value.toString().toByteArray(Charsets.UTF_8)
        override fun parse(bytes: ByteArray): JsonObject =
            SyncKitJson.instance.parseToJsonElement(bytes.toString(Charsets.UTF_8)).jsonObject
        override fun merge(local: JsonObject, remote: JsonObject): JsonObject =
            snapshot((items(local) + items(remote)).distinct().sorted(), maxOf(updatedAt(local), updatedAt(remote)))
        override fun fingerprint(value: JsonObject): String = items(value).sorted().joinToString(",")
        override fun updatedAt(value: JsonObject): String = value.getValue("updatedAt").jsonPrimitive.content
    }

    private fun items(value: JsonObject) = value.getValue("items").jsonArray.map { it.jsonPrimitive.content }
    private fun updatedAt(value: JsonObject) = value.getValue("updatedAt").jsonPrimitive.content
    private fun snapshot(items: List<String>, updatedAt: String) = buildJsonObject {
        put("items", JsonArray(items.map(::JsonPrimitive)))
        put("updatedAt", updatedAt)
    }
    private fun at(minute: Int) = "2026-09-23T00:%02d:00.000Z".format(minute)

    /** Keeps each credential's PRF secret; clearing it simulates losing the passkey. */
    private class Passkeys(private val crypto: V1EnvelopeCrypto<JsonObject>) : KeyProvider {
        val authenticator = mutableMapOf<String, ByteArray>()
        private var created = 0

        override suspend fun create(activity: Activity, appId: String): CreatedKey {
            val credentialId = "credential-${++created}"
            val secret = crypto.randomBytes(32).also { authenticator[credentialId] = it }
            val metadata = V1KeyMetadata(credentialId, "recovery.example", crypto.randomBytes(32), crypto.randomBytes(32))
            return CreatedKey(metadata, crypto.deriveContentKey(secret, metadata.kdfSalt))
        }

        override suspend fun unlock(activity: Activity, envelope: SyncEnvelopeV1): ByteArray {
            val secret = authenticator[envelope.credentialId]
                ?: throw SyncKitError(SyncKitErrorCode.KEY, "This passkey is not available on this device.")
            return crypto.deriveContentKey(secret, Base64Url.decode(envelope.kdfSalt))
        }

        override fun clear() = Unit
    }

    private class Cloud : CloudStore {
        var envelope: SyncEnvelopeV1? = null

        override suspend fun find(appId: String, authorization: Authorization): StoredEnvelope? =
            envelope?.let { StoredEnvelope("file", it) }

        override suspend fun write(
            appId: String,
            envelope: SyncEnvelopeV1,
            authorization: Authorization,
            existingId: String?,
        ): String {
            this.envelope = envelope
            return "file"
        }

        override suspend fun delete(appId: String, fileId: String, authorization: Authorization) {
            envelope = null
        }
    }

    private class Device(val controller: SnapshotSyncController<JsonObject>, val crypto: V1EnvelopeCrypto<JsonObject>) {
        var local: JsonObject? = null
    }

    private fun device(profile: V1CompatibilityProfile, cloud: Cloud, keys: Passkeys, local: JsonObject): Device {
        val crypto = V1EnvelopeCrypto(profile, codec)
        lateinit var created: Device
        val controller = SnapshotSyncController(
            SnapshotSyncOptions(
                appId = profile.appId,
                codec = codec,
                envelopeCrypto = crypto,
                keyProvider = keys,
                authorizationProvider = object : AuthorizationProvider {
                    override suspend fun authorize() = Authorization("token")
                },
                cloudStore = cloud,
                readLocal = { created.local ?: local },
                applyMerged = { created.local = it },
                activity = { Activity() },
            ),
        )
        created = Device(controller, crypto)
        return created
    }

    private inline fun rejects(message: String, block: () -> Unit): SyncKitError {
        val error = try {
            block()
            null
        } catch (error: SyncKitError) {
            error
        }
        assertNotNull("expected a SyncKitError containing \"$message\"", error)
        assertTrue("\"${error!!.message}\" does not contain \"$message\"", error.message.orEmpty().contains(message))
        return error
    }

    @Test
    fun leavesV1ExactlyAsItWasUnlessAProfileOptsIn() = runBlocking {
        val appProfile = profile()
        val cloud = Cloud()
        val phone = device(appProfile, cloud, Passkeys(V1EnvelopeCrypto(appProfile, codec)), snapshot(listOf("a"), at(1)))
        phone.controller.setup()
        val envelope = checkNotNull(cloud.envelope)
        assertEquals(1, envelope.schemaVersion)
        assertNull(envelope.appId)
        assertNull(envelope.passkeyKey)
        val encoded = SyncKitJson.instance.encodeToString(SyncEnvelopeV1.serializer(), envelope)
        assertTrue(listOf("appId", "contentSalt", "passkeyKey", "recoveryKey").none { encoded.contains("\"$it\"") })
        val error = rejects("readVersions") { phone.controller.setRecoveryCode(RecoveryCodes.generate()) }
        assertEquals(SyncKitErrorCode.CONFIGURATION, error.code)
    }

    @Test
    fun recoversALostPasskeyOnANewDeviceWithTheCodeAlone() = runBlocking {
        val appProfile = profile(readVersions = listOf(1, 2))
        val cloud = Cloud()
        val keys = Passkeys(V1EnvelopeCrypto(appProfile, codec))
        val phone = device(appProfile, cloud, keys, snapshot(listOf("a"), at(1)))
        phone.controller.setup()
        val code = RecoveryCodes.generate()
        phone.controller.setRecoveryCode(code)
        assertEquals(2, cloud.envelope?.schemaVersion)
        assertEquals("recovery-app", cloud.envelope?.appId)

        phone.local = snapshot(listOf("a", "b"), at(2))
        val lockBefore = cloud.envelope?.recoveryKey
        phone.controller.sync(SyncReason.CHANGE)
        assertEquals(2, cloud.envelope?.schemaVersion)
        assertEquals(lockBefore, cloud.envelope?.recoveryKey)

        keys.authenticator.clear()
        val newDevice = device(appProfile, cloud, keys, snapshot(listOf("c"), at(3)))
        rejects("not available") { newDevice.controller.sync(SyncReason.FOREGROUND) }
        val recovered = newDevice.controller.recover(code)
        assertEquals(SnapshotOperation.RECOVER, recovered.operation)
        assertEquals(SyncOutcome.RECOVERED, recovered.outcome)
        assertEquals(listOf("a", "b", "c"), items(checkNotNull(recovered.value)))
        assertEquals(SyncOutcome.UNCHANGED, newDevice.controller.sync(SyncReason.FOREGROUND).outcome)
        assertEquals(listOf("a", "b", "c"), items(newDevice.crypto.decryptWithRecoveryCode(checkNotNull(cloud.envelope), code)))
    }

    @Test
    fun distinguishesATypoFromAWrongCodeAndRemovesTheCode(): Unit = runBlocking {
        val appProfile = profile(readVersions = listOf(1, 2))
        val cloud = Cloud()
        val phone = device(appProfile, cloud, Passkeys(V1EnvelopeCrypto(appProfile, codec)), snapshot(listOf("a"), at(1)))
        phone.controller.setup()
        val code = RecoveryCodes.generate()
        phone.controller.setRecoveryCode(code)
        val envelope = checkNotNull(cloud.envelope)
        val typo = code.substring(0, 5) + (if (code[5] == 'A') 'B' else 'A') + code.substring(6)
        rejects("typo") { phone.crypto.decryptWithRecoveryCode(envelope, typo) }
        rejects("does not unlock") { phone.crypto.decryptWithRecoveryCode(envelope, RecoveryCodes.generate()) }
        phone.controller.setRecoveryCode(null)
        assertNull(cloud.envelope?.recoveryKey)
        rejects("no recovery code") { phone.crypto.decryptWithRecoveryCode(checkNotNull(cloud.envelope), code) }
    }

    @Test
    fun migratesExplicitlyAndReversiblyAndV1DropsTheCode() = runBlocking {
        val appProfile = profile(readVersions = listOf(1, 2))
        val cloud = Cloud()
        val keys = Passkeys(V1EnvelopeCrypto(appProfile, codec))
        val phone = device(appProfile, cloud, keys, snapshot(listOf("a"), at(1)))
        phone.controller.setup()
        phone.controller.setRecoveryCode(RecoveryCodes.generate())
        phone.controller.migrateVersion(1)
        assertEquals(1, cloud.envelope?.schemaVersion)
        assertNull(cloud.envelope?.recoveryKey)
        val old = device(profile(), cloud, keys, snapshot(emptyList(), at(0)))
        assertEquals(listOf("a"), items(checkNotNull(old.controller.sync(SyncReason.FOREGROUND).value)))
        phone.controller.migrateVersion(2)
        assertEquals(2, cloud.envelope?.schemaVersion)
    }

    @Test
    fun refusesV2WithoutOptingInAndAuthenticatesTheHeaderAndAppId(): Unit = runBlocking {
        val appProfile = profile(readVersions = listOf(1, 2), writeVersion = 2)
        val cloud = Cloud()
        val keys = Passkeys(V1EnvelopeCrypto(appProfile, codec))
        device(appProfile, cloud, keys, snapshot(listOf("a"), at(1))).controller.setup()
        val envelope = checkNotNull(cloud.envelope)
        assertEquals(2, envelope.schemaVersion)
        rejects("readVersions") { device(profile(), cloud, keys, snapshot(emptyList(), at(0))).controller.sync(SyncReason.FOREGROUND) }
        val key = keys.unlock(Activity(), envelope)
        val crypto = V1EnvelopeCrypto(appProfile, codec)
        rejects("could not be decrypted") { crypto.decrypt(envelope.copy(updatedAt = at(9)), key) }
        rejects("belongs to recovery-app") {
            V1EnvelopeCrypto(profile(appId = "another-app", readVersions = listOf(1, 2)), codec).decrypt(envelope, key)
        }
    }

    @Test
    fun validatesVersionsInTheProfile() {
        for ((reads, write) in listOf(listOf(2) to 2, listOf(1) to 2, listOf(1, 3) to 1)) {
            val error = try {
                profile(readVersions = reads, writeVersion = write)
                null
            } catch (error: IllegalArgumentException) {
                error
            }
            assertNotNull("readVersions=$reads writeVersion=$write should be rejected", error)
        }
    }

    // --- Cross-platform: the web package wrote it, Android reads it ---------

    private val fixture: JsonObject by lazy {
        val stream = checkNotNull(javaClass.classLoader?.getResourceAsStream("v2/snapshot-recovery.json")) {
            "Missing v2/snapshot-recovery.json test resource."
        }
        SyncKitJson.instance.parseToJsonElement(stream.bufferedReader().use { it.readText() }).jsonObject
    }

    private fun fixtureProfile(readVersions: List<Int> = listOf(1, 2)): V1CompatibilityProfile {
        val input = fixture.getValue("profile").jsonObject
        return profile(
            appId = input.getValue("appId").jsonPrimitive.content,
            aad = input.getValue("aad").jsonPrimitive.content,
            hkdfInfo = input.getValue("hkdfInfo").jsonPrimitive.content,
            readVersions = readVersions,
        )
    }

    private fun fixtureEnvelope(name: String): SyncEnvelopeV1 =
        V1EnvelopeCrypto(fixtureProfile(), codec).parseEnvelope(
            fixture.getValue("envelopes").jsonObject.getValue(name).toString(),
        )

    private fun fixtureKey(crypto: V1EnvelopeCrypto<JsonObject>, passkey: String): ByteArray {
        val entry = fixture.getValue("passkeys").jsonObject.getValue(passkey).jsonObject
        return crypto.deriveContentKey(
            Base64Url.decode(entry.getValue("prfSecret").jsonPrimitive.content),
            Base64Url.decode(entry.getValue("kdfSalt").jsonPrimitive.content),
        )
    }

    private fun expected(name: String): List<String> =
        items(fixture.getValue("expected").jsonObject.getValue(name).jsonObject)

    @Test
    fun decryptsEveryWebWrittenSnapshotWithItsPasskeyAndItsCode() {
        val crypto = V1EnvelopeCrypto(fixtureProfile(), codec)
        val code = fixture.getValue("recoveryCode").jsonPrimitive.content
        val original = fixtureKey(crypto, "original")
        val replacement = fixtureKey(crypto, "replacement")
        assertEquals(expected("plain"), items(crypto.decrypt(fixtureEnvelope("plain"), original)))
        assertEquals(expected("withRecovery"), items(crypto.decrypt(fixtureEnvelope("withRecovery"), original)))
        assertEquals(expected("withRecovery"), items(crypto.decryptWithRecoveryCode(fixtureEnvelope("withRecovery"), code)))
        // After recovery the snapshot opens with the new passkey and still with the code.
        assertEquals(expected("relocked"), items(crypto.decrypt(fixtureEnvelope("relocked"), replacement)))
        assertEquals(expected("relocked"), items(crypto.decryptWithRecoveryCode(fixtureEnvelope("relocked"), code)))
        rejects("could not open") { crypto.decrypt(fixtureEnvelope("relocked"), original) }
        val migrated = fixtureEnvelope("migrated")
        assertEquals(1, migrated.schemaVersion)
        assertEquals(expected("migrated"), items(V1EnvelopeCrypto(fixtureProfile(listOf(1)), codec).decrypt(migrated, original)))
    }

    @Test
    fun refusesAWebWrittenV2SnapshotWhenTheProfileReadsOnlyV1() {
        rejects("readVersions") {
            V1EnvelopeCrypto(fixtureProfile(listOf(1)), codec).parseEnvelope(
                fixture.getValue("envelopes").jsonObject.getValue("plain").toString(),
            )
        }
    }

    // --- Cross-platform: Android writes it, the web package reads it --------

    /**
     * Writes Android-built v2 snapshots for scripts/check-recovery-parity.sh,
     * which opens them with the web package — with the passkey secret and with
     * the Android-sealed recovery code. A no-op unless SNAPSHOT_RECOVERY_OUTPUT
     * is set.
     */
    @Test
    fun writesAndroidBuiltSnapshotsForTheWebToOpen() {
        val output = System.getenv("SNAPSHOT_RECOVERY_OUTPUT") ?: return
        val appProfile = profile(appId = "fixture-recovery-kotlin", readVersions = listOf(1, 2), writeVersion = 2)
        val crypto = V1EnvelopeCrypto(appProfile, codec)
        val secret = bytes(32)
        val metadata = V1KeyMetadata("credential-kotlin", "fixture.example", bytes(32), bytes(32))
        val key = crypto.deriveContentKey(secret, metadata.kdfSalt)
        val code = RecoveryCodes.generate()
        val value = snapshot(listOf("kotlin", "shared"), at(5))
        val withRecovery = crypto.setRecoveryCode(crypto.encrypt(value, key, metadata), key, code)
        val report = buildJsonObject {
            put(
                "profile",
                buildJsonObject {
                    put("appId", appProfile.appId)
                    put("aad", appProfile.aad)
                    put("hkdfInfo", appProfile.hkdfInfo)
                },
            )
            put("prfSecret", Base64Url.encode(secret))
            put("recoveryCode", code)
            put("envelope", SyncKitJson.instance.encodeToJsonElement(SyncEnvelopeV1.serializer(), withRecovery))
            put("expected", value)
        }
        File(output).writeText(report.toString())
    }
}
